### DISCUSSION TITLE
Bulk insertion not propagating default values

### DISCUSSION THREAD

**Discussion by Sinclert** [sqlalchemy/sqlalchemy#6273] [Usage Questions]
**Describe the bug**
The objects being inserted using the  [`bulk_save_objects()`](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/orm/session.py#L3397) method, when the `return_defaults` flag is set to `True`, do not have their values updated.

**Expected behavior**
Object fields that are not specified, but have a `default` value for the column, should be updated upon successful insertion.

**To Reproduce**
```python
from datetime import datetime

from sqlalchemy.engine import create_engine
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import scoped_session
from sqlalchemy.orm import sessionmaker
from sqlalchemy import Column
from sqlalchemy import DateTime
from sqlalchemy import String

from typing import List

base = declarative_base()


class DummyModel(base):
    """ Dummy data model """

    __tablename__ = "dummy"

    object_id = Column(String(32), nullable=False, primary_key=True)
    created_at = Column(DateTime, nullable=True, default=datetime.utcnow)

    @property
    def data(self) -> dict:
        return {c.name: getattr(self, c.name) for c in self.__table__.columns}


class DummyModelController:
    """ Dummy controller for the data model """

    model = DummyModel

    def __init__(self, connection_url: str):
        self.engine = create_engine(connection_url)
        self.session = scoped_session(sessionmaker(bind=self.engine))

    def setup(self):
        base.metadata.create_all(bind=self.engine)

    def get(self, obj_id: str):
        return self.session.get(self.model, obj_id)

    def create(self, obj: DummyModel):
        self.session.add(obj)
        self.session.commit()

    def create_bulk(self, objs: List[DummyModel]):
        self.session.bulk_save_objects(objs, return_defaults=True)  # Using the flag 👀 
        self.session.commit()


controller = DummyModelController("sqlite:///:memory:")
controller.setup()


# Test 1 (using regular `add`) ✅
dummy = DummyModel(object_id="example-1")
controller.create(dummy)
record = controller.get("example-1")

assert record.data == dummy.data


# Test 2 (using `bulk_save_objects`) ❌
dummy = DummyModel(object_id="example-2")
controller.create_bulk([dummy])
record = controller.get("example-2")

assert record.data == dummy.data
```

**Error**

```sh
AssertionError: assert {'created_at': datetime.datetime(2021, 4, 14, 21, 35, 28, 149804), 'object_id': 'example-2'} == {'created_at': None, 'object_id': 'example-2'}

```

**Versions.**
 - OS: mac OS 11.2.3
 - Python: Python 3.7.9
 - SQLAlchemy: 1.4.6
 - Database: SQLite
 - DBAPI: -

**Additional context**
I have tried setting the `return_defaults` flag both to `True` and to `False`. In both cases, objects that skip values for the columns using the `default` argument, do not have their internal values updated when using [`bulk_save_objects()`](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/orm/session.py#L3397).

I assumed this flag was precisely created for this purpose. Feel free to close this issue if this is not the case.

**Comment by zzzeek** [comment]
The primary key values are populated in those objects, but this is not the purpose of bulk_save_objects().    if you want other defaults to be populated you can set [eager_defaults=True](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html?highlight=eager_defaults#sqlalchemy.orm.mapper.params.eager_defaults).

**Reply by a-gwn** [reply]
I've tried with [eager_defaults=True](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html?highlight=eager_defaults#sqlalchemy.orm.mapper.params.eager_defaults) and it still didn't populate the other defaults.

If it's not the purpose of [bulk_save_objects](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.bulk_save_objects), then the documentation is a bit misleading?

It mentions:
The objects as given are not added to the session and no additional state is established on them, unless the return_defaults flag is also set, in which case primary key attributes and **server-side default values will be populated**.

**Reply by zzzeek** [reply]
that's a simple bug.  does not seem to work in 1.3 either.   But I really would prefer people dont use bulk_save_objects()

**Reply by zzzeek** [reply]
ill just update the docs to not include that

**Reply by zzzeek** [reply]
have you tried the techniques described at https://docs.sqlalchemy.org/en/14/orm/persistence_techniques.html#using-insert-update-and-on-conflict-i-e-upsert-to-return-orm-objects ?  these will be actually more efficient and provide fully stateful ORM objects with no limitations

**Reply by zzzeek** [reply]
updated docs in 4cca61c93d

**Reply by a-gwn** [reply]
wow didn't notice that, thanks a lot!
